[Upgrade Details] Implement tracking of `UPG_ROLLBACK` state

ycombinator opened the pull request:
## What does this PR do?

This PR enhances the Elastic Agent to track the `UPG_ROLLBACK` state.

## Why is it important?

When an Agent upgrade is rolled back, we want to know about it.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~I have made corresponding changes to the documentation~
- [ ] ~I have made corresponding change to the default configuration files~
- [x] ~I have added tests that prove my fix is effective or that my feature works~
- [x] ~I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~
- [ ] I have added an integration test or an E2E test

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Relates to https://github.com/elastic/elastic-agent/issues/3119#issuecomment-1743638526

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3694/detail/PR-3694/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3694/detail/PR-3694/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3694/detail/PR-3694/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3694/detail/PR-3694/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-agent_3694.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-11-02T17:46:05.109Z&rangeTo=2023-11-02T18:06:05.109Z&transactionName=BUILD+elastic-agent%2Felastic-agent-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=f1a8fd884faef89b8d1d4070ef81943c&transactionId=95b82392fd6e41b2)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-11-02T17:56:05.109+0000


* Duration: 6 min 12 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3694/detail/PR-3694/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `check`
<ul>
<li>Took 1 min 6 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/elastic-agent/pipelines/elastic-agent-mbp/pipelines/PR-3694/runs/1/steps/175/log/?start=0">here</a></li>
<li>Description: <code>make check-ci</code></l1>
</ul>
  
  </p>
  </details>



## :grey_exclamation: Flaky test report
No test was executed to be analysed.



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages.


- `run integration tests` : Run the Elastic Agent Integration tests.


- `run end-to-end tests` : Generate the packages and run the E2E Tests.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

ycombinator commented on testing/integration/upgrade_rollback_test.go:
```diff
@@ -119,6 +120,12 @@ inputs:
+	// ensure that upgrade details now show the state as UPG_ROLLBACK
+	state, err := startFixture.Client().State(ctx)
+	require.NoError(t, err)
+	require.NotNil(t, state.UpgradeDetails) // FIXME: there is no UpgradeDetails on State (yet!)
```
The `UpgradeDetails` field on `state` will become available once https://github.com/elastic/elastic-agent/pull/3615 is merged.

elasticmachine commented:
Pinging @elastic/elastic-agent (Team:Elastic-Agent)

ycombinator commented:
Unit test failure in CI is relevant to this PR:

```
=== Failed
--
  | === FAIL: internal/pkg/agent/application/upgrade TestWatchMarker (0.05s)
  | step_mark_test.go:79:
  | Error Trace:	/Users/admin/builds/bk-agent-prod-orka-1699028124327725357/elastic/elastic-agent/internal/pkg/agent/application/upgrade/step_mark_test.go:79
  | /Users/admin/.gvm/versions/go1.20.10.darwin.amd64/src/runtime/asm_amd64.s:1598
  | Error:      	Expected value not to be nil.
  | Test:       	TestWatchMarker
```

Will investigate and fix.

ycombinator commented:
Rollback-related integration tests are failing.  Most likely related to the changes in this PR.  I'm investigating...

pchila commented on internal/pkg/agent/application/upgrade/marker_watcher.go:
```diff
@@ -0,0 +1,124 @@
+		mfw.updateCh <- *marker
+	}
+
+	<-ctx.Done()
```
we launch a goroutine but then we block here... 
We have the goroutine just to start watching the file as soon as possible while we try and do the first read? 
Couldn't we refactor this to have a single go routine handle the watch and the first read? This would avoid possible races if a write happens between the start of the goroutine and the first read and both try to send the same marker on the update channel

AndersonQ replied:
I agree with Paolo, if it's gonna block anyway, I don't see why the goroutine is needed.

ycombinator replied:
Agreed, I did not implement this well.  Will refactor.  Thanks for calling it out.

ycombinator replied:
Refactored in 83577da728.

pchila commented on internal/pkg/agent/application/upgrade/marker_watcher_test.go:
```diff
@@ -0,0 +1,84 @@
+
+	go func() {
+		err := markerWatcher.Run(testCtx)
+		require.NoError(t, err)
```
Because of `t.FailNow()` called by `require`, shouldn't the `require.*` be called from the goroutine running the test function?

AndersonQ replied:
yeah, `require` does not work on a goroutine

AndersonQ replied:
if you need to abort the test, you'd need to signal the test goroutine to exit

ycombinator replied:
Addressed as part of the refactoring done in 83577da728.

pchila commented on internal/pkg/agent/application/upgrade/marker_watcher_test_util.go:
```diff
@@ -0,0 +1,30 @@
+	execPath, err := os.Executable()
+	require.NoError(t, err)
+
+	testDataDir := filepath.Join(filepath.Dir(execPath), "data")
```
Since this is only to be used in tests could you rename the file to `marker_watcher_util_test.go` or make it available only to test code in some way ?

AndersonQ replied:
+1

AndersonQ replied:
or just move it to the file where it's used

ycombinator replied:
> Since this is only to be used in tests could you rename the file to `marker_watcher_util_test.go` or make it available only to test code in some way ?

If I rename the file `marker_watcher_util_test.go` it only becomes available to test code in the same package.  However, I need this function in a test in another package too.  That's why my solution was to name it `marker_watcher_test_util.go`.  This is the same approach taken by [`logger.NewTesting`](https://github.com/elastic/elastic-agent/blob/72653ac2b0b57deedd276c3abc8d19f9085b3154/pkg/core/logger/testing.go#L15-L26).

> or just move it to the file where it's used

It's used from two different packages, so unfortunately it can't be moved to where it's used (without duplication).

pchila replied:
I would rather have the duplication and break the unit test dependency between the 2 packages...
Why would this detail be needed in 2 packages? Is there more than 1 package that needs to access the marker file directly and that dependency cannot be mocked away ?

ycombinator replied:
>  Is there more than 1 package that needs to access the marker file directly and that dependency cannot be mocked away ?

The unit test outside this package that calls his helper function is [`application.TestLimitsLog`](https://github.com/elastic/elastic-agent/blob/64563943437a4703e56498ca215a2a0881e4984b/internal/pkg/agent/application/application_test.go#L50).  If I remove the call to `upgrade.CreateMarkerTestDir(t)` from that test, it fails like so:

```
=== RUN   TestLimitsLog
    application_test.go:69: 
        	Error Trace:	/Users/shaunak/development/github/elastic-agent/internal/pkg/agent/application/application_test.go:69
        	Error:      	Received unexpected error:
        	            	failed to create upgrader: failed to set watch on upgrade marker's directory [/private/var/folders/mc/qr61t2n91z3g63ffq5sy4cm00000gn/T/GoLand/data]: lstat /private/var/folders/mc/qr61t2n91z3g63ffq5sy4cm00000gn/T/GoLand/data: no such file or directory
        	Test:       	TestLimitsLog
--- FAIL: TestLimitsLog (0.00s)
```

That's because that unit test calls `Application.New`, which ends up eventually calling `upgrade.newMarkerFileWatcher()`, which is where this error is coming from because the upgrade marker's directory doesn't exist at that point.  In normal (non-test) execution, the directory exists because it's created at installation time, before Agent is run and `Application.New` is called.

Anyway, I think we can remove this dependency if I move the instantiation of the watch inside Run() as you're alluding to in https://github.com/elastic/elastic-agent/pull/3694#discussion_r1389062050.  So let me give that a shot and see if it works.

ycombinator replied:
As expected, after simplifying the implementation of `UpgradeMarkerWatcher` in 8b9d67c797848ffc33c2c3fe39bf9751b2aded64, I was able to remove this dependency as the `UpgradeMarkerWatcher` constructor no longer adds the watch on the upgrade marker's directory; that is now being done in the `Run()` method instead, which is not in the execution path of this test.

pchila reviewed (commented):
A couple of comments mostly about concurrency and encapsulation, the change looks ok for the rest

AndersonQ commented on internal/pkg/agent/application/coordinator/coordinator.go:
```diff
@@ -1224,14 +1254,15 @@ func (c *Coordinator) filterByCapabilities(comps []component.Component) []compon
 // Called on the main Coordinator goroutine.
-func collectManagerErrors(timeout time.Duration, varsErrCh, runtimeErrCh, configErrCh chan error) error {
+func collectManagerErrors(timeout time.Duration, varsErrCh, runtimeErrCh, configErrCh, upgradeMarkerWatchErrCh chan error) error {
 	var runtimeErr error
 	var configErr error
 	var varsErr error
+	var upgradeMarkerErr error
 	// in case other components are locked up, let us time out
 	timeoutWait := time.NewTimer(timeout)
 	defer timeoutWait.Stop()
-	var returnedRuntime, returnedConfig, returnedVars bool
+	var returnedRuntime, returnedConfig, returnedVars, returnedUpgradeMarker bool
```
[NIT]
This could be more consistent

```go
	var runtimeErr, configErr, varsErr, upgradeMarkerErr error
	var returnedRuntime, returnedConfig, returnedVars, returnedUpgradeMarker bool

	// in case other components are locked up, let us time out
	timeoutWait := time.NewTimer(timeout)
	defer timeoutWait.Stop()
```

ycombinator replied:
Fixed in 1cd2e07938.

AndersonQ commented on internal/pkg/agent/application/coordinator/coordinator.go:
```diff
@@ -876,14 +890,25 @@ func (c *Coordinator) runner(ctx context.Context) error {
 		varsErrCh <- nil
 	}
 
+	upgradeMarkerWatcherErrCh := make(chan error, 1)
+	if c.upgradeMgr != nil && c.upgradeMgr.MarkerWatcher() != nil {
+		go func() {
+			err := c.upgradeMgr.MarkerWatcher().Run(ctx)
+			cancel()
+			upgradeMarkerWatcherErrCh <- err
+		}()
+	} else {
+		upgradeMarkerWatcherErrCh <- nil
+	}
+
```
The [collectManagerErrors](https://github.com/elastic/elastic-agent/pull/3694/files#diff-efef2d2536ab052feff95a0b0565325806b28e733bb9f2abc319b646194d4de5R1290-R1291) uses the error to know the upgrade marker run (`returnedUpgradeMarker`)

ycombinator replied:
Yeah, it's what @AndersonQ said.  It's the same pattern followed in the blocks of code directly above this one for the runtime manager, the config manager, and the vars manager.

ycombinator replied:
With the refactoring for https://github.com/elastic/elastic-agent/pull/3694#discussion_r1386317575, the call to `c.upgradeMgr.MarkerWatcher().Run(ctx)` will no longer return an error.  As such, this whole business with the  `upgradeMarkerWatcherErrCh` will go away.

ycombinator replied:
> With the refactoring for [#3694 (comment)](https://github.com/elastic/elastic-agent/pull/3694#discussion_r1386317575), the call to `c.upgradeMgr.MarkerWatcher().Run(ctx)` will no longer return an error. As such, this whole business with the `upgradeMarkerWatcherErrCh` will go away.

Addressed in 83577da728.

AndersonQ commented on internal/pkg/agent/application/upgrade/details/details.go:
```diff
@@ -68,6 +68,7 @@ func NewDetails(targetVersion string, initialState State, actionID string) *Deta
 
 // SetState is a convenience method to set the state of the upgrade and
 // notify all observers.
+// Do NOT call SetState with StateFailed; call the Fail method instead.
```
Why? So I'm guessing the Fail method is much more than just sugar syntax, am I right?

ycombinator replied:
That's right. The [`Fail()` method](https://github.com/elastic/elastic-agent/blob/72653ac2b0b57deedd276c3abc8d19f9085b3154/internal/pkg/agent/application/upgrade/details/details.go#L99-L117) sets some other fields besides `state` as well.  It also takes an `error` as an input parameter, otherwise I would've simply called it directly from within `SetState(s State)` if `s == StateFailed`. 

With this godoc line at least there is some warning to the user not to call `SetState(StateFailed)` but call `Fail(err)` instead.  I'm open to better ideas, though.

AndersonQ commented on internal/pkg/agent/application/upgrade/details/details.go:
```diff
@@ -118,6 +119,14 @@ func (d *Details) RegisterObserver(observer Observer) {
 }
 
+// Equals compares the non-lock fields of two Details structs.
+func (d *Details) Equals(otherD *Details) bool {
```
+1

ycombinator replied:
Yeah, good catch.  This method was originally comparing the struct values so I didn't have such checks but it makes sense to add them now.  Thanks!

ycombinator replied:
Fixed in 9171b1ada787ef3eb92b4bd5952d1e216e6c5a26.

AndersonQ commented on internal/pkg/agent/application/upgrade/marker_watcher.go:
```diff
@@ -0,0 +1,124 @@
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create upgrade marker watcher: %w", err)
+	} // Watch the upgrade marker file's directory, not the file itself, so we
+
+	// notice the file even if it's deleted and recreated.
```
[Suggestion]
I guess those are the same comment, right?
```suggestion
	} 
	
	// Watch the upgrade marker file's directory, not the file itself, so we
	// notice the file even if it's deleted and recreated.
```

ycombinator replied:
Yeah, thanks, not sure what happened here 😄.  Maybe something during a rebase + conflict resolution 🤷.

ycombinator replied:
Fixed in ea5177522c58e011b59c354ff55954ef5d69e1eb.

AndersonQ commented on internal/pkg/agent/application/upgrade/marker_watcher.go:
```diff
@@ -0,0 +1,124 @@
+			case <-ctx.Done():
+				return
+			case err, ok := <-mfw.watcher.Errors:
+				mfw.logger.Debug("after there are watch errors")
```
[Suggestion]
Could you add a wee more context to be more clear after _what_ there are watch errors?

ycombinator replied:
Actually this is a holdover from some debugging I was doing.  This line should just be deleted.

ycombinator replied:
Deleted as part of the refactoring done in https://github.com/elastic/elastic-agent/commit/83577da72817b8c9817f1cd3f7b467825a9654f8.

AndersonQ commented on internal/pkg/agent/application/upgrade/marker_watcher.go:
```diff
@@ -0,0 +1,124 @@
+				mfw.errCh <- fmt.Errorf("upgrade marker watch returned error: %w", err)
+				continue
+			case e, ok := <-mfw.watcher.Events:
+				mfw.logger.Debugf("after there is a watch event: [%s]", e.String())
```
Again, after _what_?

ycombinator replied:
The line is also holdover from some debugging I was doing. This line should just be deleted.

ycombinator replied:
Deleted as part of the refactoring done in https://github.com/elastic/elastic-agent/commit/83577da72817b8c9817f1cd3f7b467825a9654f8.

AndersonQ commented on internal/pkg/agent/application/upgrade/marker_watcher.go:
```diff
@@ -0,0 +1,124 @@
+					return
+				}
+
+				if e.Name != mfw.markerFilePath {
+					// Event is for a file other than the upgrade marker; ignore it.
+					continue
+				}
```
[Suggestion]
if it isn't expected to happen, I'd say there should be a warning log, if it's expected to happen, could you say so in the comment?

ycombinator replied:
It's expected to happen.  I will expand the comment.  Thanks.

ycombinator replied:
Done in d874790f7c.

AndersonQ commented on internal/pkg/agent/application/upgrade/marker_watcher.go:
```diff
@@ -0,0 +1,124 @@
+	// is already present before the watching starts.
+	marker, err := loadMarker(mfw.markerFilePath)
+	if err != nil {
+		return fmt.Errorf("unable to load upgrade marker from watch: %w", err)
```
+1

ycombinator replied:
When I wrote this code here I did inspect `loadMarker` to ensure that [it returned `nil, nil` if the file was not present](https://github.com/elastic/elastic-agent/blob/72653ac2b0b57deedd276c3abc8d19f9085b3154/internal/pkg/agent/application/upgrade/step_mark.go#L156-L158).  So if the file does not exist yet or is still being written to, `loadMarker` here will not return an error (which is what we want).

So the only errors that could be returned from `loadMarker` are:
- if there's some error reading the file into memory, or
- if there's some error parsing the contents of the file as YAML into memory

I think the question is whether either of these errors could be transient.  That is, if a call to `loadMarker` were to be attempted again (as part of the goroutine code), could these errors go away or will they always be returned. 

I can't come up with conditions that would make the first error transient but perhaps there are some.

As for the second error, I suppose there's some scenario in which the file is still being written to so only part of the contents have been written. So that would make the second error transient.

To address the (potentially) transient nature of these errors, I'm good with removing the error check here and just checking if the returned `marker` is `nil` (which we are doing already in the code).

ycombinator replied:
I will address this as part of the refactoring for https://github.com/elastic/elastic-agent/pull/3694#discussion_r1386317575.

ycombinator replied:
> I will address this as part of the refactoring for [#3694 (comment)](https://github.com/elastic/elastic-agent/pull/3694#discussion_r1386317575).

Addressed as part of the refactoring done in 83577da728.

AndersonQ commented on internal/pkg/agent/application/upgrade/marker_watcher.go:
```diff
@@ -0,0 +1,124 @@
+		}
+	}()
+
+	// Do an initial read from the upgrade marker file, in case the file
+	// is already present before the watching starts.
+	marker, err := loadMarker(mfw.markerFilePath)
```
if it's the first read, shouldn't it happen before the goroutine be started?

ycombinator replied:
Ideally it should happen _after_ the goroutine has started or as part of the goroutine's `select`.  If the first read were done before the goroutine started we could encounter the following undesirable scenario:
1. Upgrade marker file is read (first read).
2. Upgrade marker file gets updated.
3. Goroutine starts, but update event is now "lost".

The refactoring done in 83577da72817b8c9817f1cd3f7b467825a9654f8 ensures that the first read happens in the same goroutine `select` as the waiting on update events from the `fsnotify.Watcher`.

AndersonQ commented on internal/pkg/agent/cmd/watch.go:
```diff
@@ -107,10 +108,35 @@ func watchCmd(log *logp.Logger, cfg *configuration.Configuration) error {
 	ctx := context.Background()
 	if err := watch(ctx, tilGrace, errorCheckInterval, log); err != nil {
-		log.Error("Error detected proceeding to rollback: %v", err)
+		log.Error("Error detected, proceeding to rollback: %v", err)
```
+1

ycombinator replied:
@AndersonQ What is the "+1" here for?

ycombinator commented:
@pchila @AndersonQ Thank you for the great review.  I believe I've addressed all your feedback now.  Please re-review when you get a chance.  Thanks again!

ycombinator commented on internal/pkg/agent/application/coordinator/coordinator.go:
```diff
@@ -1225,13 +1246,13 @@ func (c *Coordinator) filterByCapabilities(comps []component.Component) []compon
-	var configErr error
-	var varsErr error
+	var runtimeErr, configErr, varsErr error
+	var returnedRuntime, returnedConfig, returnedVars bool
+
 	// in case other components are locked up, let us time out
 	timeoutWait := time.NewTimer(timeout)
 	defer timeoutWait.Stop()
-	var returnedRuntime, returnedConfig, returnedVars bool
+
```
Cleanup unrelated to the functional changes introduced in this PR.

michalpristas commented on internal/pkg/agent/application/upgrade/details/details.go:
```diff
@@ -118,6 +119,22 @@ func (d *Details) RegisterObserver(observer Observer) {
+
+	if d == nil || otherD == nil {
+		return false
+	}
```
up to you: comparing addresses first to avoid comparing metadata

ycombinator replied:
Sorry, I didn't follow what you're suggesting.

ycombinator replied:
I think you're suggesting that if the addresses are equal then we can short circuit and return true.  I'll make that change, thanks.

ycombinator replied:
Done in 96b824b74e.

michalpristas commented on internal/pkg/agent/application/upgrade/marker_watcher.go:
```diff
@@ -0,0 +1,124 @@
+				}
+
+				switch {
+				case e.Op&fsnotify.Create == fsnotify.Create ||
```
alt. compare `!=0`

ycombinator replied:
Done in 42b8a41805.

michalpristas commented on internal/pkg/agent/cmd/watch.go:
```diff
@@ -107,10 +108,35 @@ func watchCmd(log *logp.Logger, cfg *configuration.Configuration) error {
+		}
+
+		marker.Details.SetState(details.StateRollback)
+		err = upgrade.SaveMarker(marker)
```
i wonder if we want to guard this with some retry. we're accessing marker file from 2 processes now and when one process opens marker even for reading the other may fail on windows

ycombinator replied:
Thanks, I was not aware this was an issue on Windows.  So it seems it would make sense to add retry in `upgrade.SaveMarker` and `upgrade.LoadMarker` both.  I'll make the change.

ycombinator replied:
Implemented in c174487f9a.  Let me know what you think.

cmacknz replied:
That this can fail makes detecting the upgrade rollback state best effort. We might be able to improve this, and make it so that only getting the reason for a rollback is best effort.

When the agent first starts up it reads the upgrade marker to determine if it needs to acknowledge a pending upgrade action. This eventually ends up here:

https://github.com/elastic/elastic-agent/blob/f7dcbd74ce0e27d3b46b478755168c66d78dc111/internal/pkg/agent/application/upgrade/upgrade.go#L210-L236

If the agent at this point sees that there is an upgrade marker file, and the target version does not match the agent's current version, that should be enough to conclude the upgrade was rolled back. I don't think there's another situation that leads to this.

This should guarantee we can report rollbacks if this write fails for whatever reason, but we just won't know the reason why.

This may be complicated by the fact that it looks like the upgrade watcher sometimes removes the marker file itself? I'm not sure why or if this is correct because per the code above the presence of the marker is how the agent knows to ack the upgrade. This could be causing some of our "stuck in updating" problems if I am following this correctly.

https://github.com/elastic/elastic-agent/blob/f7dcbd74ce0e27d3b46b478755168c66d78dc111/internal/pkg/agent/cmd/watch.go#L118-L128

ycombinator replied:
> If the agent at this point sees that there is an upgrade marker file, and the target version does not match the agent's current version, that should be enough to conclude the upgrade was rolled back.

I think the logic here needs to be revised slightly. 

Older versions of Agent that don't contain upgrade details in the Upgrade Marker will only contain the version that Agent was being upgraded _from_ in the Upgrade Marker (the `prevVersion` field), not the target version of the upgrade.

As such, I think the condition indicating a rollback is if the `prevVersion` value from the Upgrade Marker matches the agent's current version.

---

Also, while I totally agree with this idea of doing a best effort under the said circumstances of recording in upgrade details that we've rolled back, I don't think this `Ack()` method is the right place to do it as it's specific to Fleet and also it's concern is acknowledging the upgrade.  I think we can make this logic the responsibility of the Upgrade Marker Watcher introduced in this PR as that's not a Fleet-specific component.

ycombinator replied:
Implemented in 210af5c9a5.  Please let me know what you think.

michalpristas commented on internal/pkg/agent/application/upgrade/details/details_test.go:
```diff
@@ -92,3 +92,26 @@ func TestDetailsDownloadRateJSON(t *testing.T) {
+
+	details3 := NewDetails("8.12.0", StateDownloading, "foobar")
+
+	require.True(t, details1.Equals(details2))
```
```suggestion
	require.True(t, details1.Equals(details1))
	require.True(t, details1.Equals(details2))
```

michalpristas commented on internal/pkg/agent/application/upgrade/marker_access_windows.go:
```diff
@@ -0,0 +1,66 @@
+		return os.WriteFile(markerFile, markerBytes, 0600)
+	}
+
+	if err := accessMarkerFileWithRetries(writeFn); err != nil {
```
let's see how it behaves and if there will be any risk of out of order writes. e.g retry writes after another call succeeds. for that we would need to implement some queue mechanism but that sounds like too much and it's possible this will just work

michalpristas commented on internal/pkg/agent/application/upgrade/step_download_test.go:
```diff
@@ -83,13 +83,17 @@ func TestDownloadWithRetries(t *testing.T) {
 		},
 	}
 
+	CreateMarkerTestDir(t)
```
want to do cleanup after?

ycombinator replied:
The `CreateMarkerTestDir` does the cleanup.  That's why it's passed the `t`.

michalpristas reviewed (approved):
I think it looks OK, i let others comment on whether they see their comments addressed

pchila commented on internal/pkg/agent/application/upgrade/marker_watcher.go:
```diff
@@ -0,0 +1,124 @@
+	// Watch the upgrade marker file's directory, not the file itself, so we
+	// notice the file even if it's deleted and recreated.
+	upgradeMarkerDirPath := filepath.Dir(upgradeMarkerFilePath)
+	err = watcher.Add(upgradeMarkerDirPath)
```
Why do we add the files to watch here but we instantiate the goroutine to consume events only in Run() ?

Can't we instantiate the watcher directly in `Run()` so that we can also add a 
```
defer watcher.Close()
```
on the goroutine, making sure we close the watcher ?
If that is the case maybe we should document it on the interface stating that,  whatever happens,  a created watcher must always be closed, even if we didn't call Run() or other functions on it

ycombinator replied:
Yes, it should be possible to do all this work inside `Run()`.  Let me give it a shot.

ycombinator replied:
The reason I added the files to watch in the constructor instead of `Run()` was because I only wanted the files to be added to the watch once per `MarkerFileWatcher` instance.  In theory, the calling code could call `Run()` repeatedly on the same `MarkerFileWatcher` instance, and I didn't want to add the same files to the watch each time.  I don't know how `fsnotify` behaves if the same path is watched multiple times — will it emit duplicate events? Or is it smart enough to de-dupe paths each time they're added? 

Let me play with `fsnotify` some more to see if watching the same paths is safe.  If it is, I think we can greatly simplify the implementation here.

ycombinator replied:
I experimented with `fsnotify` and confirmed that the `Add()` method is safe.  With that, I was able to simplify the implementation of the `UpgradeMarkerWatcher` quite a bit in 8b9d67c797.

pchila commented on internal/pkg/agent/application/upgrade/marker_watcher.go:
```diff
@@ -0,0 +1,123 @@
+			case err, ok := <-mfw.watcher.Errors:
+				if !ok { // Channel was closed (i.e. Watcher.Close() was called).
+					mfw.logger.Debug("fsnotify.Watcher's error channel was closed")
+					return
```
shouldn't we close the watcher when returning? Or we are relying on the caller to always do it ?

ycombinator replied:
As of the current implementation, we're relying on the caller to always do it but this could all change depending on what I find after https://github.com/elastic/elastic-agent/pull/3694#discussion_r1389177469.

ycombinator replied:
After simplifying the `UpgradeMarkerWatcher` implementation in 8b9d67c797848ffc33c2c3fe39bf9751b2aded64, I was able to move the `fsnotify` watcher's initialization and closing into the `Run()` method in 49d16e94ef.  As such, the caller doesn't need to do anything any more.

pchila commented on internal/pkg/agent/application/upgrade/marker_watcher.go:
```diff
@@ -0,0 +1,123 @@
+					// and send them over the update channel.
+					mfw.processMarker()
+				}
+			case <-doInitialRead:
```
This now makes Run() asynchronous (before we were blocking on `<- ctx.Done()`)
What I mean with my comments on the previous version of this was that since Run() was blocking we didn't need the extra goroutine :sweat_smile: .
Did the wait move somewhere else ?

ycombinator replied:
While `Run()` was blocking on `<-ctx.Done()` in the previous version, `Run()` was being called from inside a goroutine, effectively making it asynchronous.  

Now, `Run()` itself is asynchronous and I've removed the goroutine wrapper around the call to `Run()`.

pchila reviewed (commented):
Code looks ok as far as the feature is concerned.
There are a couple of clarifications needed on the MarkerWatcher usage and lifecycle and some nitpicks about package dependency (non-blocking)

pchila reviewed (approved):
LGTM

cmacknz commented on internal/pkg/agent/cmd/watch.go:
```diff
@@ -107,10 +108,35 @@ func watchCmd(log *logp.Logger, cfg *configuration.Configuration) error {
+		// because the Agent we upgraded FROM would've written upgrade details in the upgrade
+		// marker. However, if we're upgrading from version < 8.12.0, the marker won't
+		// contain upgrade details, so we populate them now.
+		if marker.Details == nil {
```
Might be worth adding a log line when this happens, in case it happens unexpectedly

ycombinator replied:
Added in fd29a3297e.

cmacknz commented on internal/pkg/agent/application/upgrade/marker_watcher.go:
```diff
@@ -0,0 +1,131 @@
+	// recorded in the upgrade marker we just read here, it would mean the
+	// upgrade was rolled back. So we explicitly create upgrade details in the
+	// upgrade marker with the UPG_ROLLBACK state.
+	if marker.Details == nil && marker.PrevVersion == version.GetAgentPackageVersion() {
```
Does this handle the case where the watcher failed to write the rolled back state? Isn't it possible for details to not be nil, but be stuck in the watching/monitoring state in this case?

ycombinator replied:
You're right.  This sufficient condition for detecting rollback here should be if the marker is present but `marker.PrevVersion == version.GetAgentPackageVersion()`.    Fixing.

ycombinator replied:
Fixed in f735acd21d.

ycombinator replied:
With added/updated unit tests: 99883dd3dc

ycombinator commented:
All integration tests involving Agent installation are failing on Windows. Investigating...

ycombinator commented:
> All integration tests involving Agent installation are failing on Windows. Investigating...

Turns out all the tests fail trying to install the Agent with the error:

```
error starting service: failed to start service (Elastic Agent): The service did not respond to the start or control request in a timely fashion.
```

I'm able to reproduce this error if I build Agent for Windows from this PR's branch and try to install it.

The only Windows-specific code in this PR is in `marker_access_windows.go`.  For debugging, I temporarily replaced this code with the same code as in `marker_access_other.go` and the install worked.  So it's definitely something about that code that's wonky.  Investigating...

cmacknz commented:
That error means the Windows service control event handler is either blocked or more likely not being started:

https://github.com/elastic/elastic-agent/blob/f7dcbd74ce0e27d3b46b478755168c66d78dc111/internal/pkg/agent/cmd/run.go#L134

My guess its the problem is in the following function getting blocked or taking too long https://github.com/elastic/elastic-agent/blob/f7dcbd74ce0e27d3b46b478755168c66d78dc111/internal/pkg/agent/cmd/run.go#L111-L113

https://github.com/elastic/elastic-agent/blob/f7dcbd74ce0e27d3b46b478755168c66d78dc111/internal/pkg/agent/cmd/run.go#L574-L582

ycombinator commented:
I've made two changes to fix the installation problems on Windows (see previous two comments for details):
- On Windows, where we implement a retry mechanism for reading the upgrade marker file, overrode the default timeout of 15 minutes to 10 seconds.
- Short circuit the retry mechanism in case the file is not found.  This is the same logic as before this PR, where we were only reading the upgrade marker once without retries in `upgrade.LoadMarker`.

All integration tests involving Agent installation are now passing on Windows.
